File: leetcode/19-remove-Nth-node-from-end-of-list/main.go
package main

import (
	"fmt"
)

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	start := &ListNode{0, head}

	slow, fast := start, start
	for i := 0; i < n; i++ {
		fast = fast.Next
	}

	for fast.Next != nil {
		fast = fast.Next
		slow = slow.Next
	}
	slow.Next = slow.Next.Next

	return start.Next
}

func main() {
	list := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 4,
			Next: &ListNode{
				Val: 5,
				Next: &ListNode{
					Val: 2,
					Next: &ListNode{
						Val: 10,
					},
				},
			},
		},
	}
	fmt.Println(removeNthFromEnd(list, 3))
}

File: leetcode/19-remove-Nth-node-from-end-of-list/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x, next = None):
        self.val = x
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0, head)
        first, second = dummy, dummy
        for i in range(n):
            first = first.next
        while first.next:
            second = second.next
            first = first.next
        second.next = second.next.next

        return dummy.next


solution = Solution()
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
print(solution.removeNthFromEnd(head, 4))
